### PR TITLE
Consider named parameters during rename refactorings

### PR DESCRIPTION
**Pull request by mlangc**
This PR implements support for named parameters when doing rename refactorings (see [Ticket #1002501](https://www.assembla.com/spaces/scala-ide/tickets/1002501-rename-does-not-consider-named-parameters/details#), as well as [Ticket #1002572](https://www.assembla.com/spaces/scala-ide/tickets/1002572-rename-refactoring-for-case-class-parameter-should-also-rename-named-arguments-to-apply-and-copy/details#)). To achieve this, this PR takes care of several things:

1. Named parameters are desugared. Therefore we have to manually parse them in the original source code. Unfortunately, due to the compiler setting incorrect positions on some of the trees in the AST, this is harder than it should be. The parsing is done with the `SourceWithMarker` framework, that now supports backtracking (note that I've strong evidence for the fact that the newly added backtracking support should not cause performance regressions: https://github.com/mlangc/scala-refactoring-experiments/blob/master/src/test/scala/com/github/mlangc/experiments/refactoring/SourceWithMarkerBenchmark.scala).
2. Named parameters are often used in connection with constructors and case classes. In these cases, there is a one-to-one mapping between certain named parameters and class values (think of constructors and generated `copy` and `apply` methods). This is what the updates related to `SymbolExpander` are about.
3. Renaming a method parameter is no longer a local operation in the sense of `Rename.PreparationResult.hasLocalScope` if the method might be called from outside the file that defines it.
4. The performance problems that led to support for named arguments being dropped in the past should be addressed by 168d91fd22b85a65d2930dd2bdd7224d55377f14. I did some benchmarking (see https://github.com/mlangc/scala-refactoring-experiments/blob/master/src/test/scala/com/github/mlangc/experiments/refactoring/RenameBenchmark.scala) and did not see any changes in the results, but there might be corner cases that are not covered by the benchmark.

Note that it is currently not possible to rename named parameters where they are used, but only where they are defined, that is you cannot initiate a rename refactoring at `f(tryRenameMe = 22)`, but only at `def f(tryRenameMe: Int)`. In my opinion, we should take care of this in a separate PR though because

- this is not a regression
- this PR is quite large already
- fixing this issue would also require changes to Scala-IDE and possibly ENSIME

### CODE REVIEW

**Comment by misto** [comment]
IIRC, one of the reasons the ApplyExtracted and BlockExtracter were commented out was that it had a huge impact on performance. Do you have any idea how the current situation is? We unfortunetly don't have tests for this.

**Comment by mlangc** [comment]
@misto The performance problems you are referring to should be addressed by a6a7f859dacbfa47d651b64383b8af203d26bd93.

**Comment by mlangc** [comment]
Darn, I forgot to test with Scala-2.10...

File: src/main/scala/scala/tools/refactoring/implementations/Rename.scala [hunk 1]
```diff
@@ -29,22 +29,73 @@ abstract class Rename extends MultiStageRefactoring with TreeAnalysis with analy
 
   def prepare(s: Selection) = {
 
+    /*
+     * This heuristic decides if this refactoring is local to a single file or not. Note that
```
Comment (misto) [inline src/main/scala/scala/tools/refactoring/implementations/Rename.scala]: Looks like a sensible trade off between safety and ease of use.

File: src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala [hunk 2]
```diff
@@ -95,6 +94,107 @@ trait DependentSymbolExpanders extends TracingImpl {
...
+      case _ => None
+    }
+
+    private def findRelatedCtorParamSymbolIn(parent: Tree, valSym: Symbol): Option[Symbol] = {
+      parent.foreach {
+        case dd: DefDef if dd.symbol.isConstructor =>
+          val res = findParamSymbolAssociatedWithValSymbol(dd, valSym)
+
+          if (res.nonEmpty) {
+            return res
```
Comment (misto) [inline src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala]: Nitpicking here: I'm not a fan of exlicit returns, but I see why you're using it here.

**Comment by misto** [comment]
I played around with a few of the tests but couldn't break it. Nice job! Unfortunately, there's a minor conflict (with the recent imports removal Simon did) so we can't merge it automatically.

File: src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala (hunk 2 above)
Comment (sschaef) [inline src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala]: `parent.iterator.collect{ case dd: DefDef if dd.symbol.isConstructor => findParamSymbolAssociatedWithValSymbol(dd, valSym)}.find(_.nonEmpty)` should do the same

Comment (misto) [inline src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala]: Right, I forgot there's an interator to make the computation lazy..

Comment (mlangc) [inline src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala]: @sschaef that's an elegant solution indeed, but `Tree` does not have an `iterator` method.

**Comment by mlangc** [comment]
@misto I've just rebased.

File: src/test/scala/scala/tools/refactoring/tests/util/SourceWithMarkerTest.scala [hunk 3]
```diff
@@ -306,9 +352,152 @@ class SourceWithMarkerTest {
...
+    }
+
+    {
+      val src = SourceWithMarker("aaaa")
+      assertTrue(src.moveMarker(('a' || 'b').zeroOrMore).isDepleted)
+    }
+
+    {
+      val src = SourceWithMarker("aaaa")
+      assertTrue(src.moveMarker('a'.zeroOrMore.zeroOrMore).isDepleted)
```
Comment (wpopielarski) [inline src/test/scala/scala/tools/refactoring/tests/util/SourceWithMarkerTest.scala]: it's not a problem with this PR but maybe in future it's worth to consider renaming `isDepleted` to `isConsumed`. Currently it looks akin to `PimpedTree`.

File: src/main/scala/scala/tools/refactoring/util/SourceWithMarker.scala [hunk 4]
```diff
@@ -274,16 +335,16 @@ object SourceWithMarker {
   }
 
   object Movement {
-    def apply(impl: (SourceWithMarker, Boolean) => Option[Int]): Movement = {
+    def apply(impl: (SourceWithMarker, Boolean) => Seq[Int]): Movement = {
```
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/util/SourceWithMarker.scala]: would be fine to write what `apply` takes as argument in `trait Movement` (especially what is in returned `Seq` / `Option` ) (better Scaladoc)

File: src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala (hunk 2 above)
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala]: not sure about that `return`. will this guy throw exception? I guess. Maybe we need early return traverser for `Tree` (akin to `FindTraverser` or `CollectTraverser`)?

File: src/main/scala/scala/tools/refactoring/common/TreeTraverser.scala [hunk 5]
```diff
@@ -25,11 +25,68 @@ trait TreeTraverser {
...
+
+        val stackTrace = Thread.currentThread().getStackTrace.drop(3).take(5)
+        stackTrace.map("  -- " + stackElemToString(_)).mkString("\n")
+      }
+
+      if (indent == 0) {
+        context(s"Starting traversal in $sourceFileAsString at $treeAsString initiated from\n$explainStack\n") {
+          val start = System.currentTimeMillis()
+          doTraverse()
+          val stop = System.currentTimeMillis()
```
Comment (sschaef) [inline src/main/scala/scala/tools/refactoring/common/TreeTraverser.scala]: Difference between two calls of `System.nanoTime` divided by `1e6` results in more exact timings + handles timings < 1ms

File: src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala (hunk 2 above)
Comment (mlangc) [inline src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala]: Yes, nonlocal returns are implemented with exceptions under the hood: http://stackoverflow.com/questions/17704050/how-does-scala-implement-return-from-within-an-expression. If we had `collectFirst` for trees, this hack wouldn't be needed and the code definitely clearer. I'm considering this...

File: src/test/scala/scala/tools/refactoring/tests/util/SourceWithMarkerTest.scala (hunk 3 above)
Comment (mlangc) [inline src/test/scala/scala/tools/refactoring/tests/util/SourceWithMarkerTest.scala]: I agree... `isConsumed` sounds better indeed. A rename refacoring might be in order ;-). However, this might cause some conflicts with PR #144, so maybe we should wait with this until both PRs are merged?

File: src/main/scala/scala/tools/refactoring/util/SourceWithMarker.scala (hunk 4 above)
Comment (mlangc) [inline src/main/scala/scala/tools/refactoring/util/SourceWithMarker.scala]: I've just added Scaladocs to this method.

File: src/main/scala/scala/tools/refactoring/common/TreeTraverser.scala (hunk 5 above)
Comment (mlangc) [inline src/main/scala/scala/tools/refactoring/common/TreeTraverser.scala]: Well, a lot has been written about this, and there are pros and cons, but I agree with you.

**Comment by sschaef** [comment]
Impressive work. Worked very well in my tests. LGTM

File: src/test/scala/scala/tools/refactoring/tests/util/SourceWithMarkerTest.scala (hunk 3 above)
Comment (misto) [inline src/test/scala/scala/tools/refactoring/tests/util/SourceWithMarkerTest.scala]: +1 for a separate PR so we can merge this one

**Comment by wpopielarski** [comment]
Generally LGTM, but I have one general question about refactoring. There is `ApplyExtractor` and `BlockExtractor` and other and all refactoring actions should use them or every refactoring action should rather define its own well tailored extractors? I think it is a design question. For example I have a problem now in organize imports action concerning the discovering of usages stable identifiers from `import global._`. As I understand it is done by `CompilationUnitDependencies` and the big issue at least for novice is to add some snippet of code into it with fear that some other functionality can be broken.

File: src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala (hunk 2 above)
Comment (wpopielarski) [inline src/main/scala/scala/tools/refactoring/analysis/SymbolExpanders.scala]: good idea with `collectFirst`. Traverser would looks like this ?:
```scala
{
var collected = None
override def traverse(t: Tree) = t match {
  if (collected.nonEmpty) return
  if (collected.isEmpty && pf.isDefinedAt(t)) collected = pf(t)
  super.traverse(t)
}
```

**Comment by sschaef** [comment]
I think both `ApplyExtractor` and `BlockExtractor` miss better Scaladoc that explains exactly what they are doing. These non-scalac tree nodes introduce additional complexity that potentially needs to be faced everywhere.
@wpopielarski I didn't understand what exactly your problem is but we can talk later about it.

**Comment by mlangc** [comment]
@wpopielarski I see your point: Both `ApplyExtractor` and `BlockExtractor` have been there before this PR, I just enhanced them. The only traverser that uses `ApplyExtractor` explicitly is `TreeWithSymbolTraverse`. Apart from that, there is also `scala.tools.refactoring.common.EnrichedTrees.children`, that nobody seems to use according to `References -> Workspace` (I'm not sure if this is reliable). `BlockExtractor` is not used in any traverser explicitly, but indirectly via `TemplateExtractor`. So you are right, the situation is quite confusing, and it would be great to bring some order into this.

As for `CompilationUnitDependencies`: This is certainly not the easiest part of the library, but it should not be affected by the changes to `ApplyExtractor` or `BlockExtractor`. Also, there are quite a few tests for this module. If you break it, the chances that one of them fail are high, so if you want to change something there, just do it and see what happens ;-).

**Comment by mlangc** [comment]
@sschaef I just added some Scaladocs for `ApplyExtractor`. I wanted to do the same for `BlockExtractor` but realized, that I don't understand it well enough myself (it wasn't added in this PR and wrong documentation is worse than no documentation). 

What puzzles me for example, is that both `ApplyExtractor` _and_ `BlockExtractor` concern themselves with named arguments. I tried to get rid of the (limited) named arguments handling in `BlockExtractor` but ran into various failing tests (mostly related to tree printing). 

@misto, maybe you could shine some light on this?

**Comment by misto** [comment]
What I attempted to do with the various XyExtractors was to sanitize or undo some of the desugaring the compiler does, so that during pattern matching over a tree, we get NamedArgument trees even though they are not really part of the AST. And IIRC, depending on how named arguments calls are ordered, they are sometimes desugared to Blocks and sometimes simply to Apply trees (I believe when the named parameters are passed in the same order as the defined parameters, it's an Apply tree). The story is similar with multiple-assignments (```val (a, b) = ..```), which is actually a pattern match that introduces a lot of additional trees and weird positions..

**Comment by misto** [comment]
And yeah, luckily we have a very high test coverage, so if you manage to simplify something and get rid of code and no test breaks, then just go for it.

**Comment by mlangc** [comment]
@misto Thanks for the clarification; I've added a comment to `BlockExtractor` that highlights that it is mostly (in fact I think only - but didn't prove it) needed for tree printing.

**Comment by sschaef** [comment]
Let's test this in production
